krm-functions/lib/kptfile/v1/kpt.go

```diff
@@ -0,0 +1,167 @@
+/*
+Copyright 2023 Nephio.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1
+
+import (
+	"errors"
+
+	"github.com/GoogleContainerTools/kpt-functions-sdk/go/fn"
+	kptv1 "github.com/GoogleContainerTools/kpt/pkg/api/kptfile/v1"
+	"sigs.k8s.io/yaml"
+)
+
+type KptFile interface {
+	// Unmarshal decodes the raw document within the in byte slice and assigns decoded values into the out value.
+	// it leverages the  "sigs.k8s.io/yaml" library
+	UnMarshal() (*kptv1.KptFile, error)
+	// Marshal serializes the value provided into a YAML document based on "sigs.k8s.io/yaml".
+	// The structure of the generated document will reflect the structure of the value itself.
+	Marshal() ([]byte, error)
+	// ParseKubeObject returns a fn sdk KubeObject; if something failed an error
+	// is returned
+	ParseKubeObject() (*fn.KubeObject, error)
+	// GetKptFile returns the Kptfile as a go struct
+	GetKptFile() *kptv1.KptFile
+	// SetConditions sets the conditions in the kptfile. It either updates the entry if it exists
+	// or appends the entry if it does not exist.
+	SetConditions(...kptv1.Condition)
+	// DeleteCondition deletes the condition equal to the conditionType if it exists
+	DeleteCondition(ct string)
+	// GetCondition returns the condition for the given ConditionType if it exists,
+	// otherwise returns nil
+	GetCondition(ct string) *kptv1.Condition
+	// GetConditions returns all the conditions in the kptfile. if not initialized it
+	// returns an emoty slice
+	GetConditions() []kptv1.Condition
+}
+
+// NewMutator creates a new mutator for the kptfile
+// It expects a raw byte slice as input representing the serialized yaml file
+func NewMutator(b string) KptFile {
+	return &kptFile{
+		raw: []byte(b),
+	}
```

Comment:
My suggestion is to create the complete  object here by marshaling the byte array and make the state of the object usable immedietly

Reply:
I changed the name to New here as the code also has the namespace.
I did not like the NewMUtator name, but here is the challenge. For this library we only operate on kptfile and add/delete conditions and access information of the resource.

Now for ip allocation we have 2 different operations. 1 we have to create an allocation from scratch and 2 access and modify the data. I believe there need to be 2 different structs that handle this.

So I believe the one that create it I want to call NewGenerator and the other one New. So in IP Allocation we need both. A generator and a reader/parser and updater. This is my dilemma in general. For this we are ok with just the reader/parser and updater